src/IR.jl

```diff
@@ -0,0 +1,1127 @@
+    handle::API.MlirDialectHandle
+end
+
+function DialectHandle(s::Symbol)
+    s = Symbol("mlirGetDialectHandle__", s, "__")
+    DialectHandle(getproperty(API, s)())
+end
+
+Base.convert(::Type{MlirDialectHandle}, handle::DialectHandle) = handle.handle
+
+### Dialect Registry
+
+mutable struct DialectRegistry
+    registry::MlirDialectRegistry
+end
+function DialectRegistry()
+    registry = API.mlirDialectRegistryCreate()
+    @assert !mlirIsNull(registry) "cannot create DialectRegistry with null MlirDialectRegistry"
+    finalizer(DialectRegistry(registry)) do registry
+        API.mlirDialectRegistryDestroy(registry.registry)
+    end
+end
+
+function Base.insert!(registry::DialectRegistry, handle::DialectHandle)
+    API.mlirDialectHandleInsertDialect(registry, handle)
+end
+
+### Context
+
+mutable struct Context
+    context::MlirContext
+end
+function Context()
+    context = API.mlirContextCreate()
+    @assert !mlirIsNull(context) "cannot create Context with null MlirContext"
+    finalizer(Context(context)) do context
+        API.mlirContextDestroy(context.context)
+    end
+end
+
+Base.convert(::Type{MlirContext}, c::Context) = c.context
+
+num_loaded_dialects(context) = API.mlirContextGetNumLoadedDialects(context)
+function get_or_load_dialect!(context, handle::DialectHandle)
+    mlir_dialect = API.mlirDialectHandleLoadDialect(handle, context)
+    if mlirIsNull(mlir_dialect)
+        error("could not load dialect from handle $handle")
+    else
+        Dialect(mlir_dialect)
+    end
+end
+function get_or_load_dialect!(context, dialect::String)
+    get_or_load_dialect!(context, DialectHandle(Symbol(dialect)))
+end
+
+function enable_multithreading!(context, enable=true)
+    API.mlirContextEnableMultithreading(context, enable)
+    context
+end
+
+is_registered_operation(context, opname) = API.mlirContextIsRegisteredOperation(context, opname)
+
+### Location
+
+struct Location
+    location::MlirLocation
+
+    Location(location) = begin
+        @assert !mlirIsNull(location) "cannot create Location with null MlirLocation"
+        new(location)
+    end
+end
+
+Location(context::Context) = Location(API.mlirLocationUnknownGet(context))
+Location(context::Context, filename, line, column=0) =
+    Location(API.mlirLocationFileLineColGet(context, filename, line, column))
+Location(context::Context, lin::Core.LineInfoNode) =
+    Location(context, string(lin.file), lin.line)
+Location(context::Context, lin::LineNumberNode) =
+    isnothing(lin.file) ?
+    Location(context) :
+    Location(context, string(lin.file), lin.line)
+Location(context::Context, ::Nothing) = Location(context)
+
+Base.convert(::Type{MlirLocation}, location::Location) = location.location
+
+function Base.show(io::IO, location::Location)
+    c_print_callback = @cfunction(print_callback, Cvoid, (MlirStringRef, Any))
+    ref = Ref(io)
+    print(io, "Location(#= ")
+    GC.@preserve ref API.mlirLocationPrint(location, c_print_callback, ref)
```

Comment:
```suggestion
    API.mlirLocationPrint(location, c_print_callback, ref)
```

The `GC.@preserve` should not be necessary